jxl/src/image.rs

```diff
@@ -0,0 +1,257 @@
+            size: self.size,
+            data,
+        })
+    }
+}
+
+impl<'a, T: ImageDataType> ImageRectMut<'a, T> {
+    pub fn rect(
+        &'a mut self,
+        origin: (usize, usize),
+        size: (usize, usize),
+    ) -> Result<ImageRectMut<'a, T>> {
+        rect_size_check(origin, size, self.size)?;
+        Ok(ImageRectMut {
+            origin: (origin.0 + self.origin.0, origin.1 + self.origin.1),
+            size,
+            image: self.image,
+        })
+    }
+
+    pub fn size(&self) -> (usize, usize) {
+        self.size
+    }
+
+    pub fn row(&mut self, row: usize) -> &mut [T] {
+        debug_assert!(row < self.size.1);
+        let start = (row + self.origin.0) * self.image.size.1 + self.origin.1;
+        &mut self.image.data[start..start + self.size.0]
+    }
+
+    pub fn as_rect(&'a self) -> ImageRect<'a, T> {
+        ImageRect {
+            origin: self.origin,
+            size: self.size,
+            image: self.image,
+        }
+    }
+}
+
+#[cfg(feature = "debug_tools")]
+pub mod debug_tools {
+    use super::{ImageDataType, ImageRect};
+    pub trait ToU8ForWriting {
+        fn to_u8_for_writing(self) -> u8;
+    }
+
+    impl ToU8ForWriting for u8 {
+        fn to_u8_for_writing(self) -> u8 {
+            self
+        }
+    }
+
+    impl ToU8ForWriting for u16 {
+        fn to_u8_for_writing(self) -> u8 {
+            (self as u32 * 255 / 65535) as u8
+        }
+    }
+
+    impl ToU8ForWriting for f32 {
+        fn to_u8_for_writing(self) -> u8 {
+            (self * 255.0).clamp(0.0, 255.0).round() as u8
+        }
+    }
+
+    impl ToU8ForWriting for u32 {
+        fn to_u8_for_writing(self) -> u8 {
+            (self as f32 / (2.0f32.powi(32) - 1.0)).to_u8_for_writing()
+        }
+    }
+
+    impl ToU8ForWriting for half::f16 {
+        fn to_u8_for_writing(self) -> u8 {
+            self.to_f32().to_u8_for_writing()
+        }
+    }
+
+    impl<'a, T: ImageDataType + ToU8ForWriting> ImageRect<'a, T> {
+        pub fn to_pgm(&self) -> Vec<u8> {
+            use std::io::Write;
+            let mut ret = vec![];
+            write!(&mut ret, "P5\n{} {}\n255\n", self.size.0, self.size.1).unwrap();
+            ret.extend(
+                (0..self.size.1)
+                    .flat_map(|x| self.row(x).iter())
+                    .map(|x| x.to_u8_for_writing()),
+            );
+            ret
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::error::Result;
+
+    use super::Image;
+
+    #[test]
+    fn huge_image() {
+        assert!(Image::<u8>::new(1 << 28, 1 << 28).is_err());
+    }
+
+    #[test]
+    fn rect_basic() -> Result<()> {
+        let mut image = Image::<u8>::new(32, 32)?;
+        assert_eq!(image.as_rect_mut().rect((0, 0), (1, 1))?.size(), (1, 1));
+        assert!(image.as_rect_mut().rect((30, 30), (3, 3)).is_err());
+        image.as_rect_mut().rect((30, 30), (1, 1))?.row(0)[0] = 1;
+        assert_eq!(image.as_rect_mut().row(30)[30], 1);
```

Comment:
not super important, but perhaps non testing non-square images as well might make sense?